pkg/topology/kademlia/kademlia.go

```diff
@@ -231,43 +231,35 @@ func (k *Kad) connectBalanced(wg *sync.WaitGroup, peerConnChan chan<- *peerConnI
 		}
 
 		binPeers := k.knownPeers.BinPeers(uint8(i))
+		binConnectedPeers := k.connectedPeers.BinPeers(uint8(i))
 
 		for j := range k.commonBinPrefixes[i] {
 			pseudoAddr := k.commonBinPrefixes[i][j]
 
-			closestConnectedPeer, err := closestPeer(k.connectedPeers, pseudoAddr, noopSanctionedPeerFn)
+			// Connect to closest known peer which we haven't tried connecting to recently.
+
+			_, err := nClosePeerInSlice(binConnectedPeers, pseudoAddr, noopSanctionedPeerFn, uint8(i+k.bitSuffixLength+1))
 			if err != nil {
-				if errors.Is(err, topology.ErrNotFound) {
-					break
+				if !errors.Is(err, topology.ErrNotFound) {
+					return
```

Comment:
there was not a return here, why the change?

Reply:
this function does not return any other error than ErrNotFound, so this is non-consequential, I can remove this

Reply:
that's incorrect. there are 2 levels of `for` loops here, the `break` statement means that the immediate enclosing for loop breaks, but the enclosing loop continues

Reply:
the break was in a different context. 
closestpeer to any pseudoaddress always existed if there was at least some connections in the bins.
Now, closeenoughPeer can return a close-enough-peer if there is one, or nothing. 
if there's a close enough connected peer, we skip to next pseudoaddress (continue).
if there is not one, we select one in line 250. 
if there is not one, we get the topology.ErrNotFound error. if any other error is received, probably something is off, so return. this can never happen with the current nClosePeer function

Reply:
interactive review please